src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -566,39 +463,35 @@ end = struct
             ; ("ignored_hash", Root_hash.to_yojson root_hash) ]
           "My desired root was $desired_hash, so I'm ignoring $ignored_hash" ;
         () )
+      else if already_done then
+        (* This can happen if we asked for hashes that turn out to be equal in
+           underlying ledger and the target. *)
+        Logger.debug t.logger ~module_:__MODULE__ ~location:__LOC__
+          "Got sync response when we're already finished syncing"
       else
         let res =
           match (query, answer) with
-          | Query.What_hash addr, Answer.Has_hash h -> (
-            match add_child_hash_to t addr h with
+          | Query.What_child_hashes addr, Answer.Child_hashes_are (lh, rh) -> (
+            match add_child_hashes_to t addr lh rh with
             (* TODO #435: Stick this in a log, punish the sender *)
-            | Error e ->
+            | `Hash_mismatch ->
                 Logger.faulty_peer t.logger ~module_:__MODULE__
                   ~location:__LOC__
                   ~metadata:
-                    [ ("child_hash", Hash.to_yojson h)
+                    [ ("left_child_hash", Hash.to_yojson lh)
+                    ; ("right_child_hash", Hash.to_yojson rh)
                     ; ( "sender"
                       , Envelope.Sender.to_yojson
                           (Envelope.Incoming.sender env) ) ]
-                  "Got error from when trying to add child_hash $child_hash \
-                   %s $sender"
-                  (Error.to_string_hum e)
-            | Ok (`Good children_to_verify) ->
+                  "Peer lied when trying to add child hashes $left_child_hash \
+                   and $right_child_hash. Sender was: $sender"
+            | `Good children_to_verify ->
                 (* TODO #312: Make sure we don't write too much *)
                 List.iter children_to_verify ~f:(fun (addr, hash) ->
-                    handle_node t addr hash )
-            | Ok `More -> () (* wait for the other answer to come in *)
-            | Ok `Hash_mismatch ->
-                (* just ask again for both children of the parent? this is the
-                 only case where we can't immediately pin blame on a single
-                 node. *)
-                failwith "figure out how to handle peers lying" )
+                    handle_node t addr hash ) )
           | Query.What_contents addr, Answer.Contents_are leaves ->
               (* FIXME untrusted _exn *)
-              let subtree_hash =
-                add_content t addr leaves |> Or_error.ok_exn
-              in
-              Valid.set t.validity addr (Fresh, subtree_hash) |> ignore
+              add_content t addr leaves |> Or_error.ok_exn |> ignore
           | Query.Num_accounts, Answer.Num_accounts (count, content_root) ->
               handle_num_accounts t count content_root
           | query, answer ->
```

Comment:
I realize that we can probably use GADTs to no nonmatching query-answer pairs.

You do not have change the queries and answers into a GADT for this PR, but it's nice to note.

Reply:
We could force the correct correspondence of request and response types by having each of these things be a separate rpc, but I don't think there's a clean way of doing that and passing them all into the central rpc handler thing.